mlir/lib/Dialect/Vector/Transforms/VectorUnroll.cpp

```diff
@@ -1003,6 +1003,153 @@ struct UnrollFromElements : OpRewritePattern<vector::FromElementsOp> {
+///     %r0 = vector.insert_strided_slice %t3, %zero
+///       {offsets = [0, 0], strides = [1, 1]} : vector<2x2xf32> into
+///       vector<4x4xf32>
+///
+struct UnrollShapeCastPattern : public OpRewritePattern<vector::ShapeCastOp> {
+  UnrollShapeCastPattern(MLIRContext *context,
+                         const vector::UnrollVectorOptions &options,
+                         PatternBenefit benefit = 1)
+      : OpRewritePattern<vector::ShapeCastOp>(context, benefit),
+        options(options) {}
+
+  LogicalResult matchAndRewrite(vector::ShapeCastOp shapeCastOp,
+                                PatternRewriter &rewriter) const override {
+    auto targetShape = getTargetShape(options, shapeCastOp);
+    if (!targetShape)
+      return failure();
+
+    Location loc = shapeCastOp.getLoc();
+    VectorType sourceType = shapeCastOp.getSourceVectorType();
+    VectorType resultType = shapeCastOp.getResultVectorType();
+
+    ArrayRef<int64_t> resultShape = resultType.getShape();
+    ArrayRef<int64_t> sourceShape = sourceType.getShape();
+
+    SmallVector<int64_t> strides(targetShape->size(), 1);
+    Value result = rewriter.create<arith::ConstantOp>(
+        loc, resultType, rewriter.getZeroAttr(resultType));
+
+    // For each unrolled tile in the result
+    for (SmallVector<int64_t> tileOffsets :
+         StaticTileOffsetRange(resultShape, *targetShape)) {
+
+      // Create the target tile type
+      VectorType tileType =
+          VectorType::get(*targetShape, resultType.getElementType());
+
+      // Build the tile by extracting individual elements
+      Value tile = createTileFromElements(
+          rewriter, loc, shapeCastOp.getSource(), sourceShape, resultShape,
+          tileOffsets, *targetShape, tileType);
+
+      // Insert the tile into the result
+      result = rewriter.create<vector::InsertStridedSliceOp>(
+          loc, tile, result, tileOffsets, strides);
+    }
+
+    rewriter.replaceOp(shapeCastOp, result);
+    return success();
+  }
+
+private:
+  /// Creates a result tile by extracting individual elements from the source
+  /// and inserting them at the correct positions in the tile.
+  Value createTileFromElements(PatternRewriter &rewriter, Location loc,
+                               Value source, ArrayRef<int64_t> sourceShape,
+                               ArrayRef<int64_t> resultShape,
+                               ArrayRef<int64_t> tileOffsets,
+                               ArrayRef<int64_t> tileShape,
+                               VectorType tileType) const {
+
+    // Initialize tile with zeros
+    Value tile = rewriter.create<arith::ConstantOp>(
+        loc, tileType, rewriter.getZeroAttr(tileType));
+
+    // Calculate strides for both source and result shapes
+    SmallVector<int64_t> sourceStrides = computeStrides(sourceShape);
+    SmallVector<int64_t> resultStrides = computeStrides(resultShape);
+
+    // Iterate over all positions in the tile using linear indexing
```

Comment:
There are a bunch of existing helpers for (de)linearizing indices -- could we use those? For example here: https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Utils/IndexingUtils.h